### PR TITLE
CEVA: add ceva platform xc5 xm6 support

### PR DESCRIPTION
**Pull request by GUIDINGLI**
## Summary

CEVA: add ceva platform xc5 xm6 support

which code base from: https://github.com/FishsemiCode/nuttx

## Impact

New arch platform

### CODE REVIEW

File: arch/ceva/include/arch.h [hunk 1]
```diff
@@ -0,0 +1,122 @@
+/****************************************************************************
+ * arch/ceva/include/arch.h
+ *
+ *   Copyright (C) 2018 Pinecone Inc. All rights reserved.
+ *   Author: Xiang Xiao <xiaoxiang@pinecone.net>
```
Comment (xiaoxiang781216) [inline arch/ceva/include/arch.h:5]: copyright need update to apache.

**Review by xiaoxiang781216 — APPROVED** [review]

File: arch/ceva/Kconfig [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+	bool "Use D-Cache"
+	default n
+	select ARCH_DCACHE
+
+config ARCH_NR_FPUS
+	int "FPU Number"
+	default 1
+	depends on ARCH_HAVE_FPU
+	---help---
+		Build in support for the number of Floating Point Unit.
```
Comment (gustavonihei) [inline arch/ceva/Kconfig:56]:
```suggestion
		Built-in support for the number of Floating Point Unit.
```

File: arch/ceva/Kconfig [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+		Check your chip specifications first; not all CEVA chips
+		support the FPU.
+
+config ARCH_MPU
+	bool "MPU support"
+	default n
+	depends on ARCH_HAVE_MPU
+	select ARCH_USE_MPU
+	---help---
+		Build in support for the CEVA Memory Protection Unit (MPU).
```
Comment (gustavonihei) [inline arch/ceva/Kconfig:66]:
```suggestion
		Built-in support for the CEVA Memory Protection Unit (MPU).
```

File: arch/ceva/Kconfig [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+config ARCH_HAVE_HARDFAULT_DEBUG
+	bool
+	default n
+
+config DEBUG_HARDFAULT
+	bool "Verbose Hard-Fault Debug"
+	default n
+	depends on ARCH_HAVE_HARDFAULT_DEBUG && DEBUG_ALERT
+	---help---
+		Enables verbose debug output when a hard fault is occurs.  This verbose
```
Comment (gustavonihei) [inline arch/ceva/Kconfig:122]:
```suggestion
		Enables verbose debug output when a hard fault occurs.  This verbose
```

File: arch/ceva/include/arch.h [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+/* This file should never be included directed but, rather,
```
Comment (gustavonihei) [inline arch/ceva/include/arch.h:21]:
```suggestion
/* This file should never be included directly but, rather,
```

File: arch/ceva/include/arch.h [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+#  define CONFIG_ARCH_INTERRUPTSTACK CONFIG_IDLETHREAD_STACKSIZE
+#endif
+
+#if defined(CONFIG_ARCH_ADDRENV) && defined(CONFIG_BUILD_KERNEL)
+/* In the kernel build, there a multiple user heaps; one for each task
+ * group.  In this build configuration, the user heap structure lies
+ * in a reserved region at the beginning of the .bss/.data address
+ * space (CONFIG_ARCH_DATA_VBASE).  The size of that region is given by
+ * ARCH_DATA_RESERVE_SIZE
+ */
```
Comment (gustavonihei) [inline arch/ceva/include/arch.h:52]:
```suggestion
/* In the kernel build, there are multiple user heaps; one for each task
 * group.  In this build configuration, the user heap structure lies
 * in a reserved region at the beginning of the .bss/.data address
 * space (CONFIG_ARCH_DATA_VBASE).  The size of that region is given by
 * ARCH_DATA_RESERVE_SIZE
 */
```

File: arch/ceva/include/xm6/irq.h [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+#define REG_A3              REG_R3
+#define REG_A4              REG_R4
+#define REG_A5              REG_R5
+#define REG_A6              REG_R6
+#define REG_FP              REG_R8
+#define REG_LR              REG_RETREG
+#define REG_PC              REG_RETREGI
+#define REG_OM              REG_MODC  /* Operation Mode */
+
+/* MODA: IRQ enable/disable */
```
Comment (acassis) [inline arch/ceva/include/xm6/irq.h:68]: Please include a blank line after a comment line

File: arch/ceva/include/xm6/irq.h [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+#define REG_MODA_DEFAULT    0x07f0
+
+#define REG_MODA_ENABLE     0x07f0
+#define REG_MODA_DISABLE    0x0010
+
+/* MODC: Operation mode */
```
Comment (acassis) [inline arch/ceva/include/xm6/irq.h:74]: Ditto

File: arch/ceva/include/xm6/irq.h [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+#define REG_OM_DEFAULT      0x20      /* PI and Supervisor */
+
+/* Note: this is POM field not OM field */
```
Comment (acassis) [inline arch/ceva/include/xm6/irq.h:77]: Ditto

File: arch/ceva/src/common/up_assert.c [hunk 10]
```diff
@@ -0,0 +1,372 @@
...
+      up_saveusercontext(s_last_regs);
+      regs = s_last_regs;
+    }
+
+  /* Dump the interrupt registers */
+
+  for (rx = 0; rx < XCPTCONTEXT_REGS; rx += 8)
+    {
+      _alert("R%03d: %08x %08x %08x %08x %08x %08x %08x %08x\n",
+            rx, regs[rx], regs[rx + 1], regs[rx + 2], regs[rx + 3],
```
Comment (acassis) [inline arch/ceva/src/common/up_assert.c:146]: The variables line need to move a single character to the right to be aligned after the "("

File: arch/ceva/include/stdarg.h [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+#include_next <stdarg.h>
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+/* Redefine va_copy since:
+ * 1.TL4 doesn't define it at all
+ * 2.XM6/X2 generate the wrong code
+ */
```
Comment (acassis) [inline arch/ceva/include/stdarg.h:37]: Add empty line after comment lines

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by davids5** [comment]
Would a README with links to information (mfg, datasheets) add value here?

**Comment by xiaoxiang781216** [comment]
Here is some introduction info:
https://www.ceva-dsp.com/product/ceva-xm6/
https://www.ceva-dsp.com/product/ceva-xc16/
Basically, these DSPs are designed for Vision and Modem.

File: arch/ceva/include/spinlock.h [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+#  include <arch/xm6/spinlock.h>
+#endif
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+/* Spinlock states */
+
+#define SP_UNLOCKED 0  /* The Un-locked state */
```
Comment (acassis) [inline arch/ceva/include/spinlock.h:44]:
```suggestion
#define SP_UNLOCKED 0  /* The Unlocked state */
```

File: arch/ceva/include/spinlock.h [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+
+/****************************************************************************
+ * Public Types
+ ****************************************************************************/
+
+#ifndef __ASSEMBLY__
+
+/* The Type of a spinlock. */
+
+typedef uint32_t spinlock_t;
```
Comment (acassis) [inline arch/ceva/include/spinlock.h:75]: Why to redefine spinlock_t ?

File: arch/ceva/include/types.h [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+
+typedef signed short       _int16_t;
+typedef unsigned short     _uint16_t;
+
+typedef signed int         _int32_t;
+typedef unsigned int       _uint32_t;
+
+typedef signed long long   _int64_t;
+typedef unsigned long long _uint64_t;
+#define __INT64_DEFINED
```
Comment (acassis) [inline arch/ceva/include/types.h:64]: Why not to use these type from stdint.h ?

File: arch/ceva/include/spinlock.h (hunk 13 above)
Comment (Ouss4) [reply arch/ceva/include/spinlock.h:75]: This actually is the first definition of `spinlock_t`.  The one in the in `nuttx/spinlock.h` is just a stub to reduce the amount of `ifdef`s.

Comment (acassis) [reply arch/ceva/include/spinlock.h:75]: Got it! Thank you!

File: arch/ceva/include/types.h (hunk 14 above)
Comment (xiaoxiang781216) [reply arch/ceva/include/types.h:64]: stdint.h typedef come from here to handle the arch difference.

Comment (acassis) [reply arch/ceva/include/types.h:64]: Ok, I missed that, thank you!

**Comment by acassis** [comment]
@GUIDINGLI I just noticed that this CEVA arch still missing some board inclusion to use it and catch some issues that could be introduced into it. Please consider adding some (low cost) board powered by XM6 and XC16

**Comment by xiaoxiang781216** [comment]
it's hard since CEVA never release the free toolchain as far as I know.

**Pull request merged**